pkg/graveler/committed/apply.go

```diff
@@ -14,205 +14,296 @@ type ApplyOptions struct {
+	opts                  *ApplyOptions
+	summary               graveler.DiffSummary
+	haveDiffs, haveSource bool
 }
 
-// ReferenceType represents the type of the reference
-
-// applyFromSource applies all changes from source to writer.
-func applyFromSource(ctx context.Context, logger logging.Logger, writer MetaRangeWriter, source Iterator) error {
+// applyAll applies all changes from Iterator to writer and returns the number of writes
+func (a *applier) applyAll(iter Iterator) (int, error) {
+	var count int
 	for {
 		select {
-		case <-ctx.Done():
-			return ctx.Err()
+		case <-a.ctx.Done():
+			return 0, a.ctx.Err()
 		default:
 		}
-
-		sourceValue, sourceRange := source.Value()
-		if sourceValue == nil {
-			if logger.IsTracing() {
-				logger.WithFields(logging.Fields{
-					"from": string(sourceRange.MinKey),
-					"to":   string(sourceRange.MaxKey),
-					"ID":   sourceRange.ID,
-				}).Trace("copy entire source range at end")
-			}
-			if err := writer.WriteRange(*sourceRange); err != nil {
-				return fmt.Errorf("copy source range %s: %w", sourceRange.ID, err)
+		iterValue, iterRange := iter.Value()
+		if iterValue == nil {
+			if iterRange.IsTombstone() {
+				// internal error but no data lost: deletion requested of a
+				// file that was not there.
+				a.logger.WithFields(logging.Fields{
+					"from": string(iterRange.MinKey),
+					"to":   string(iterRange.MaxKey),
+					"ID":   string(iterRange.ID),
+				}).Warn("[I] unmatched delete")
+			} else {
+				if a.logger.IsTracing() {
+					a.logger.WithFields(logging.Fields{
+						"from": string(iterRange.MinKey),
+						"to":   string(iterRange.MaxKey),
+						"ID":   iterRange.ID,
+					}).Trace("copy entire range at end")
+				}
+				if err := a.writer.WriteRange(*iterRange); err != nil {
+					return 0, fmt.Errorf("copy iter range %s: %w", iterRange.ID, err)
+				}
+				count += int(iterRange.Count)
 			}
-			if !source.NextRange() {
+			if !iter.NextRange() {
 				break
 			}
 		} else {
-			if logger.IsTracing() {
-				logger.WithFields(logging.Fields{
-					"key": string(sourceValue.Key),
-					"ID":  string(sourceValue.Identity),
-				}).Trace("write key from source at end")
+			if a.logger.IsTracing() {
+				if iterValue.IsTombstone() {
+					// internal error but no data lost: deletion requested of a
+					// file that was not there.
+					a.logger.WithField("id", string(iterValue.Identity)).Warn("[I] unmatched delete")
+					continue
+				}
+				a.logger.WithFields(logging.Fields{
+					"key": string(iterValue.Key),
+					"ID":  string(iterValue.Identity),
+				}).Trace("write key from iter at end")
 			}
-			if err := writer.WriteRecord(*sourceValue); err != nil {
-				return fmt.Errorf("write source record: %w", err)
+			if err := a.writer.WriteRecord(*iterValue); err != nil {
+				return 0, fmt.Errorf("write iter record: %w", err)
 			}
-			if !source.Next() {
+			count++
+			if !iter.Next() {
 				break
 			}
 		}
 	}
-	return source.Err()
+	return count, iter.Err()
```

Comment:
E.g. after calling this method the caller has to increment the correct statistics field -- even though it sort-of has access to the statistics (it just doesn't know what type to increment).